### PR TITLE
[SPARK-46834][SQL][Collations] Support for aggregates

### PR DESCRIPTION
**Pull request by dbatomic**
### What changes were proposed in this pull request?

This PR is part of collation effort. For details please refer to umbrella JIRA ticket with linked designed document.

This PR adds support for Aggregates and GROUP BY against strings with collation.
This PR adds following:
1) `SortAggregate` support for columns with collation.
2) Rule that will force SortAggs in case of non-binary collations.
3) Tests.
4) Golden file for collations (which includes basic filters and aggs).

### Why are the changes needed?

Please refer to document in JIRA for overall collation effort.

### Does this PR introduce _any_ user-facing change?

With this PR users will have proper support for aggregates against strings that are collated with non-binary collations (e.g. UTF8_BINARY_LCASE).

### How was this patch tested?

Tests in `CollationSuite` plus golden file tests.

### Was this patch authored or co-authored using generative AI tooling?

Yes.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 1]
```diff
@@ -183,6 +186,57 @@ class CollationSuite extends DatasourceV2SQLBase {
     }
   }
 
+  test("aggregates count respects collation") {
```
Comment (stefankandic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:189]: maybe we should also test struct columns with collations inside

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 2]
```diff
@@ -405,11 +405,21 @@ abstract class HashExpression[E] extends Expression {
...
+     ctx: CodegenContext, input: String, result: String, stringType: StringType): String = {
+    if (stringType.isDefaultCollation) {
+      val baseObject = s"$input.getBaseObject()"
+      val baseOffset = s"$input.getBaseOffset()"
+      val numBytes = s"$input.numBytes()"
+      s"$result = $hasherClassName.hashUnsafeBytes($baseObject, $baseOffset, $numBytes, $result);"
+    } else {
+      val stringHash = ctx.freshName("stringHash")
+      s"""
+        long $stringHash = CollationFactory.fetchCollation(${stringType.collationId})
```
Comment (stefankandic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: why is this needed if we're going to force sort aggregate anyways?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: same question here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala [hunk 3]
```diff
@@ -353,9 +353,20 @@ object MergeScalarSubqueries extends Rule[LogicalPlan] {
         case a: AggregateExpression => a
       })
     }
+    val groupByExpressionSeq = Seq(newPlan, cachedPlan).map { plan =>
+      plan.groupingExpressions.flatMap(_.collect {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala:357]: I think it should be just `plan.groupingExpressions`, e.g. `GROUO BY func(a)` can be supported by hash aggregate if `func(a)` returns normal string, no matter if `a` is string with collation or not.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 2 above)
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: Right. I will leave this change for another PR.

We need this to make reshuffle working properly btw, but I will make that change in a separate PR.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 1 above)
Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:188]: Added test case to golden file. Please take a look.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala (hunk 3 above)
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala]: Ups, my bad. Good catch! Let me see if I can add a test that verifies this logic.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala [hunk 4]
```diff
@@ -173,7 +173,10 @@ abstract class HashMapGenerator(
             ${hashBytes(bytes)}
           """
         }
-      case _: StringType => hashBytes(s"$input.getBytes()")
+      case StringType => hashBytes(s"$input.getBytes()")
+      case st: StringType =>
+        hashLong(s"CollationFactory.fetchCollation(${st.collationId})" +
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala:178]: is this needed by this PR?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 5]
```diff
@@ -95,6 +96,20 @@ public static boolean isMutable(DataType dt) {
...
   }
 
+  /**
+   * True if comparisons, equality and hashing can be done purely on binary representation of
+   * Unsafe row. i.e. binary(e1) = binary(e2) <=> e1 = e2.
+   * e.g. this is not true for non-binary collations (any case/accent insensitive collation
+   * can lead to rows being semantically equal even though their binary representations differ).
+   */
+  public static boolean isBinaryStable(DataType dt) {
+    if (dt instanceof StringType st) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:106]: shall we check nested field/array elements recursively?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala (hunk 3 above)
Comment (dbatomic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala]: Added test please take a look.

I also added string with collation to `dsl/package` since that was a natural way to enable the optimizer tests and i guess that we need this anyhow, although this is not strictly related to aggs.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 5 above)
Comment (dbatomic) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:106]: oh, I see. Yes, we need that as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala (hunk 4 above)
Comment (dbatomic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala:178]: It is. The thing is that we are still going to use HashAgg for binary collations (e.g. UNICODE_CI which is also covered by tests). We need to do proper hashing in this case.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 5 above)
Comment (dbatomic) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:106]: done. I also moved method to `UnsafeRowUtils`. It feels a bit more natural to do this kind of checks in scala.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 6]
```diff
@@ -30,6 +30,7 @@
 import org.apache.spark.SparkUnsupportedOperationException;
 import org.apache.spark.sql.catalyst.InternalRow;
 import org.apache.spark.sql.catalyst.types.*;
+import org.apache.spark.sql.catalyst.util.CollationFactory;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java:33]: unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala [hunk 7]
```diff
@@ -197,4 +197,21 @@ object UnsafeRowUtils {
...
+  /**
+   * True if comparisons, equality and hashing can be done purely on binary representation of
+   * Unsafe row. i.e. binary(e1) = binary(e2) <=> e1 = e2.
+   * e.g. this is not true for non-binary collations (any case/accent insensitive collation
+   * can lead to rows being semantically equal even though their binary representations differ).
+   */
+  def isBinaryStable(dataType: DataType): Boolean = dataType match {
+    case st: StringType => CollationFactory.fetchCollation(st.collationId).isBinaryCollation
+    case _: AtomicType => true
+    case _: ArrayType => isBinaryStable(dataType.asInstanceOf[ArrayType].elementType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:210]: nit: `case ArrayType(et) => isBinaryStable(et)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala [hunk 8]
```diff
@@ -197,4 +197,21 @@ object UnsafeRowUtils {
...
+    case _: MapType => isBinaryStable(dataType.asInstanceOf[MapType].keyType) &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:211]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala [hunk 9]
```diff
@@ -197,4 +197,21 @@ object UnsafeRowUtils {
...
+      isBinaryStable(dataType.asInstanceOf[MapType].valueType)
+    case _: StructType =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:213]: ditto, `case StructType(fields) => ...`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala:178]: I see. I feel the code can be more readable if we do
```
case st: StringType =>
  if (st.isDefaultCollation) {

  } else {

  }
```

**Review by cloud-fan — APPROVED** [review]
LGTM except for some minor comments

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala (hunk 4 above)
Comment (dbatomic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala:178]: I did some refactoring here. Please take a look. I kept pattern matching but I think it looks a bit nicer now.

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 10]
```diff
@@ -33,6 +33,14 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
    */
   def isDefaultCollation: Boolean = collationId == StringType.DEFAULT_COLLATION_ID
 
+  /**
+   * Binary collation implies that strings are considered equal only if they are
+   * byte for byte equal. E.g. all accent or case-insensitive collations are considered non-binary.
+   * If this field is true, byte level operations can be used against this datatype (e.g. for
+   * equality and hashing).
+   */
+  def isBinaryCollation: Boolean = CollationFactory.fetchCollation(collationId).isBinaryCollation
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:42]: when shall we treat default collation and binary collation differently?

**Review by cloud-fan — APPROVED** [review]

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala (hunk 10 above)
Comment (dbatomic) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:42]: Binary collation means that you can do hash and equal (and contains) on binary representation.
Default collation means that you can do hash, equal and *sort* on binary representation.

Example of binary collation is UNICODE at identical ICU level. You need to consult ICU tables to do ordering but hashing/eq should work on byte level.

I dislike naming `default collation` and I think that we should switch to something more descriptive for UCS_BASIC. After joins, my next PR will be to:
1) Rename `UCS_BASIC` -> `UTF8_BINARY` and `UCS_BASIC_LCASE` to `UTF_BINARY_LCASE`.
2) Figure out default collation story. E.g. it should probably be something like "isBinarySortable" or something like that.

But this PR is about aggregates so let's leave this for the next PR :)

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 11]
```diff
@@ -183,6 +185,57 @@ class CollationSuite extends DatasourceV2SQLBase {
...
+    Seq(
+      ("ucs_basic", Seq("AAA", "aaa"), Seq(Row(1, "AAA"), Row(1, "aaa"))),
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:190]: This test case failed in the daily test of Maven + Java 21, both on Linux and MacOS. @dbatomic Do you have time to investigate the cause of this failure?

- linux: https://github.com/apache/spark/actions/runs/8189363924/job/22416511124

image


- macos-14: https://github.com/apache/spark/actions/runs/8194082205/job/22416483724

image

also cc @HyukjinKwon because I noticed that you are paying attention to the tests of maven + Java 21 on macos-14

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:190]: cc @dongjoon-hyun too

Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:190]: Looking into this. Thanks.

Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:190]: @stefankandic picked up fixing the issue. This is a race condition in ICU library (problem on spark side where we don't properly setup the library).

Comment (stefankandic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:190]: https://github.com/apache/spark/pull/45436
